src/graph/LookupExecutor.cpp

```diff
@@ -0,0 +1,499 @@
+                prop = *aExpr->prop();
+                filters_.emplace_back(std::make_pair(prop, rExpr->op()));
+            } else {
+                return Status::SyntaxError("Unsupported syntax ：%s", rExpr->toString().c_str());
+            }
+            break;
+        }
+        case nebula::Expression::kFunctionCall : {
+            return Status::SyntaxError("Unsupported function call ： %s",
+                                       expr->toString().c_str());
+        }
+        default : {
+            return Status::SyntaxError("Unsupported syntax ： %s", expr->toString().c_str());
+        }
+    }
+    return Status::OK();
+}
+
+Status LookupExecutor::checkFilter() {
+    auto status = traversalExpr(whereWrapper_->getFilter());
+    return status;
+}
+
+Status LookupExecutor::chooseIndex() {
+    auto ret = (isEdge_) ? ectx()->getMetaClient()->getEdgeIndexesFromCache(spaceId_)
+                         : ectx()->getMetaClient()->getTagIndexesFromCache(spaceId_);
+    if (!ret.ok()) {
+        return Status::IndexNotFound();
+    }
+
+    auto indexes = ret.value();
+    if (indexes.empty()) {
+        return Status::IndexNotFound();
+    }
+
+    auto validIndexes = findValidIndex(indexes);
+    if (validIndexes.empty()) {
+        LOG(ERROR) << "No matching index was found";
+        return Status::Error("No matching index was found");
+    }
+    return findOptimalIndex(validIndexes);
+}
+
+/**
+ * Choose the index that filter best matches, for example :
+ * the filter is col1 == 1 and col2 == 2 and col3 == 3,
+ * the indexes are index1 (col4, col3, col2, col1)
+ *                 index2 (col1, col2, col3, col4)
+ *                 index3 (col1, colA, col2, col3)
+ * as above, these indexes all are valid, but the index2 should be choose.
+ */
+Status LookupExecutor::findOptimalIndex(std::vector<IndexID> ids) {
+    Status status = Status::OK();
+    if (ids.size() == 1) {
+        index_ = ids[0];
+        return status;
+    }
+    /**
+     * Returns the first index if no optimization is required
+     */
+    if (skipOptimize_) {
+        index_ = ids[0];
+        return status;
+    }
+    auto *mc = ectx()->getMetaClient();
+
+    /**
+     * std::map<column_hint_count, IndexID>
+     */
+    std::map<int32_t, IndexID> indexHint;
+    for (auto& indexId : ids) {
+        auto itemStatus = (isEdge_) ? mc->getEdgeIndexFromCache(spaceId_, indexId)
+                                    : mc->getTagIndexFromCache(spaceId_, indexId);
+        if (!itemStatus.ok()) {
+            return itemStatus.status();
+        }
+        const auto& index = itemStatus.value();
+        int32_t hintCount = 0;
+        for (const auto& field : index->get_fields()) {
+            auto it = std::find_if(filters_.begin(), filters_.end(),
+                                   [field](const auto &rel) {
+                                       return field.get_name() == rel.first &&
```

Comment:
Only EQ could use the index?

Reply:
> Only EQ could use the index?

No , The purpose of this method is to choose an optimal index , for example : 
Suppose the where clause is `col1 == 1 and col2 == 2 and col3 == 3`,
Indexes are all correct and available as below, here can be choose best one as index2 : 
```
index1 (col4, col3, col2, col1)
index2 (col1, col2, col3, col4)
index3 (col1, colA, col2, col3)
```

Reply:
why you think index2 is the best,  the policy should be based on the data size for each index.

Reply:
> why you think index2 is the best, the policy should be based on the data size for each index.

index1's prefix : ""
index2's prefix : "col1+col2+col3"
index3.s prefix : "col1"

So the index2 is the best index.